### PR TITLE
release: backport bug fixes for 2.6.1

### PR DESCRIPTION
**Pull request by tstanczyk95**
## Summary

Prepare the Trackers 2.6.1 patch release from `release/latest`.

- Base: `release/latest` at `0e839f3`, exactly tag `2.6.0`.
- Head: `d930dcd` on `release/2.6.1`.
- Review window: `develop` was reviewed chronologically through #599 (`3fb83d1`).
- Included: 30 release commits derived from 28 upstream PRs plus 2 standalone maintenance commits.
- Scope: applicable bug fixes, release-state documentation, dependency updates, and low-risk maintenance; no new product features or performance-only changes.
- Freeze: #601–#604 arrived after the validated #599 cut and are intentionally deferred.

This remains a draft for scope and release-note review. `/release prepare 2.6.1` has not been run, so package metadata and locally built artifacts still report `2.6.0`.

## User-facing release notes

### Fixed

- **OC-SORT output (#566):** Low-confidence detections are returned with `tracker_id=-1` instead of being silently dropped. `update()` now returns one row per input detection, as documented. Applications that relied on the old filtered output can filter detections before calling `update()` or filter returned rows.
- **McByte mask cleanup (#568):** Removed-tracklet notifications remain queued until the mask pipeline receives them, including across skipped frames, duplicate timestamps, and recoverable CUDA out-of-memory failures. Tracks pruned early during timestamp-based updates are included.
- **BoT-SORT and McByte Kalman noise (#567):** Prediction refreshes process noise and measurement updates refresh measurement noise, avoiding recalculation of a matrix that the current operation does not consume.
- **Box conversion (#557):** Batched `xcycsr_to_xyxy` decoding now uses floating-point output for integer input, preserving fractional coordinates. Floating-point input retains its precision.
- **Image-sequence input (#558):** Image-directory frames now use natural filename ordering, so names such as `2.jpg` appear before `10.jpg`.
- **GIoU on malformed boxes (#589):** Box extents are clamped to zero for area calculation, preventing inverted boxes from receiving a spurious positive GIoU bonus. Scores for well-formed boxes are unchanged.
- **Documentation URLs (#582):** Generated canonical, Open Graph, structured-data, and sitemap URLs now target the deployed `/latest/` documentation alias.

### Documentation and maintenance

- Synchronize McByte benchmark tables, tracker defaults, quickstart guidance, dataset details, and README terminology with the 2.6.1 implementation.
- Improve search-result titles and add poster frames for documentation videos without backporting the post-2.6 documentation reorganization.
- Bring applicable dependency, GitHub Actions, and pre-commit pins through the agreed #599 freeze point while preserving the 2.6 CLI dependency graph.

## Included change inventory

### Functional fixes

| PR   | Source    | Release commit | Change                                                        | Application                                                |
| ---- | --------- | -------------- | ------------------------------------------------------------- | ---------------------------------------------------------- |
| #566 | `64065f1` | `eba0531`      | Return low-confidence OC-SORT detections with `tracker_id=-1` | `-x`; conflicts resolved against 2.6 docs/tests            |
| #568 | `c099233` | `f8287da`      | Make McByte mask-removal notifications durable                | `-x`; tracker conflict resolved                            |
| #567 | `da81c00` | `96b32af`      | Split Kalman Q/R refresh by consumer                          | `-x`; timestamp-test conflict resolved                     |
| #557 | `35dd78b` | `0f5a41e`      | Decode batched `xcycsr_to_xyxy` in floating point             | Direct `-x` cherry-pick                                    |
| #582 | `5864906` | `256f971`      | Emit canonical documentation URLs under `/latest/`            | `-x`; retained the exact fix without parent-only redirects |
| #558 | `8de9ca0` | `cd7bff0`      | Sort image-directory frames naturally                         | Direct `-x` cherry-pick                                    |
| #589 | `8edb175` | `ddb78c4`      | Clamp malformed-box extents for GIoU                          | `-x`; omitted parent-only test API absent from 2.6         |

### Documentation and release-alignment changes

| PR   | Source    | Release commit | Change                                                            | Application                                               |
| ---- | --------- | -------------- | ----------------------------------------------------------------- | --------------------------------------------------------- |
| #539 | `cf72d2d` | `ad60d1f`      | Add McByte default metrics and update comparison title            | Direct `-x` cherry-pick                                   |
| #544 | `b7d0f1f` | `224e2d9`      | Disable mdformat wrapping and apply resulting formatting          | Adapted during conflict resolution                        |
| #547 | `03d9e0e` | `46cf12f`      | Correct tracker docs and trim duplicated guidance                 | Adapted to the 2.6 CLI/docs layout                        |
| #555 | `54888a4` | `89bab82`      | Add McByte to summary benchmark tables                            | Adapted to files present in 2.6                           |
| #564 | `92a7940` | `458b79b`      | Fix quickstart, benchmark, detector, dataset, and license details | Adapted to 2.6 commands and paths                         |
| #583 | `ced34f0` | `48e6dce`      | Improve SERP titles and homepage targeting                        | Adapted to `docs/learn` layout                            |
| #587 | `1cb7a07` | `66c5e79`      | Describe the project as object tracking in Python                 | Adapted while retaining the existing 2.6 introduction     |
| #585 | `ca81fa5` | `de5845a`      | Add poster frames for documentation videos                        | All assets included; page paths mapped to 2.6 equivalents |

### Dependencies, CI, and repository maintenance

| PR   | Source    | Release commit | Change                                                                      | Application                                    |
| ---- | --------- | -------------- | --------------------------------------------------------------------------- | ---------------------------------------------- |
| #537 | `3180d84` | `7ba29a9`      | Bump `cryptography` 48.0.1 → 50.0.0                                         | Direct `-x` cherry-pick                        |
| #542 | `2bc14c7` | `031cbc0`      | Bump `pymdown-extensions` 11.0 → 11.0.1                                     | Direct `-x` cherry-pick                        |
| #546 | `e1cb5d3` | `223288d`      | Apply applicable grouped lock updates                                       | Regenerated for the 2.6 dependency graph       |
| #550 | `df245e0` | `4c438ca`      | Bump the shared Ruff pre-commit hook to 0.16.2                              | Adapted to the single 2.6 Ruff entry           |
| #570 | `cf39fbf` | `e4ebc7d`      | Apply the next grouped lock update                                          | Direct `-x` cherry-pick                        |
| —    | `60b21c8` | `1770d65`      | Change pre-commit autoupdates from weekly to monthly                        | Direct `-x` cherry-pick                        |
| —    | `2c9d9f2` | `330a9ee`      | Ignore local `benchmarks/` output                                           | Direct `-x` cherry-pick                        |
| #575 | `dd1cca4` | `2dece84`      | Bump `hydra-core` 1.3.2 → 1.3.4                                             | Direct `-x` cherry-pick                        |
| #581 | `40c3199` | `5310792`      | Apply applicable grouped lock updates                                       | Regenerated for the 2.6 dependency graph       |
| #588 | `665ab9f` | `f91c617`      | Apply five grouped lock updates                                             | Regenerated while preserving the 2.6 CLI state |
| #590 | `baaa7dc` | `50ccd81`      | Bump `setup-uv` and `action-gh-release`                                     | Direct `-x` cherry-pick                        |
| #591 | `f881cd4` | `ade0644`      | Bump `transformers` 5.6.2 → 5.10.1                                          | Direct `-x` cherry-pick                        |
| #593 | `8f1b4d5` | `265dc7a`      | Bump Ruff to 0.16.6 and mypy to 2.3.1                                       | Adapted to the single 2.6 Ruff entry           |
| #592 | `c25b234` | `d84b652`      | Raise the detection-extra floor and apply the final grouped lock resolution | Regenerated for the 2.6 dependency graph       |
| #599 | `3fb83d1` | `d930dcd`      | Bump Optuna, uv, build, and types-requests                                  | Direct `-x` cherry-pick; release freeze point  |

## Adapted-backport notes

The adapted commits deliberately preserve 2.6 behavior and layout instead of importing parent changes that are absent from the release branch:

- **#544:** Kept `docs/learn/cli-migration.md` deleted because that page documents the newer CLI that is not part of 2.6.1; applied the configuration change and formatting to the remaining applicable files.
- **#546:** Regenerated the 2.6 lock with the five applicable updates; omitted `jsonargparse`, which belongs to the skipped newer CLI dependency graph.
- **#547:** Omitted the absent `docs/learn/inspect.md`; retained the 2.6 `--detections`/`--mot-output` flow, McByte `cmc_downscale=2`, and corrected `--fixed-params` spelling and evaluation paths.
- **#550:** Updated one combined Ruff hook because #545, which splits Ruff into two entries, is not included.
- **#555:** Updated the homepage, README, and Copilot summary tables; did not create absent `docs/AGENTS.md` or `docs/evaluations/results.md` solely to carry cross-reference edits.
- **#564:** Applied the substantive quickstart, benchmark, detector, dataset, C-BIoU, and license corrections while retaining the 2.6 CLI commands and `docs/learn` layout; omitted the upstream 2.6.0 announcement.
- **#581:** Applied the compatible lock updates and upstream-matching transitives; omitted `jsonargparse` because the newer CLI is absent.
- **#583:** Applied equivalent page titles in the 2.6 layout; no `docs/AGENTS.md` exclusion was needed because that page is absent.
- **#588:** Applied the five upstream dependency updates while preserving the branch-specific 2.6 CLI dependency state.
- **#587:** Applied the upstream tagline and Python object-tracking definition without importing the broader README rewrite from #555.
- **#585:** Included all 17 WebP assets and updated all 12 equivalent Markdown pages. Seven `docs/guides`/`docs/evaluations` paths were mapped to their existing `docs/learn`/tracker equivalents; the skipped docs reorganization itself was not backported.
- **#593:** Updated the single combined Ruff hook and mypy; upstream has two Ruff entries only because of skipped #545.
- **#592:** Raised `inference-models` to `>=0.36.0` and applied the final upstream-matching lock resolution; omitted `jsonargparse`, retained `pydeprecate==0.11.0`, and preserved the 2.6 CLI graph.

Conflict resolutions for #566, #568, #567, #582, and #589 retained their source PRs' semantic changes while excluding unrelated parent-only content. Direct cherry-picks retain source SHAs through `cherry-pick -x`; manual adaptations identify their source commit in the commit message.

## Explicit exclusions and deferments

| Changes                                        | Decision and reason                                                                                                                                                                                                         |
| ---------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| #413, #541, #543, #552                         | New/refactored CLI surface and features; outside patch-release scope and absent from the 2.6 architecture.                                                                                                                  |
| #538                                           | Performance/default change (`cmc_downscale=6`), not a bug fix; 2.6.1 retains its implemented default of `2`.                                                                                                                |
| #545                                           | Broad docformatter/tooling migration touching 76 files, including the absent newer CLI; unnecessary for the patch and high-divergence.                                                                                      |
| #548, #549, #551, #553                         | Documentation for the newer CLI/site organization. #551 produced extensive conflicts and would import a structural reorganization rather than documentation that reflects the 2.6.1 tree.                                   |
| #565                                           | Internal agent-guidance consolidation tied to absent `AGENTS.md`, `guides`, and `evaluations` structure; no runtime or user-facing release benefit.                                                                         |
| #569                                           | Explicitly skipped after maintainer consultation. Five of six paths belong to the post-2.6 CLI; inclusion would require manually reimplementing the frame-rate fix across the older CLI/demo/benchmark code.                |
| #572, #573, #574, #576, #577, #578, #579, #580 | Performance-only changes; intentionally outside this bug-fix release.                                                                                                                                                       |
| #596                                           | Large tracklet refactor (and based partly on skipped performance work), with no required bug fix and disproportionate regression/conflict risk.                                                                             |
| #601, #602, #603, #604                         | Post-freeze Dependabot lock-only bumps (Mako, AnyIO, Accelerate, Markdown). They arrived after the fully validated #599 cut and are deferred to avoid reopening the dependency/test matrix without a 2.6.1 functional need. |

## Verification

### Local validation

Environment: macOS Apple Silicon; locked dependencies via `uv`; primary suite on Python 3.12.14 plus Python 3.10 release-compatibility checks.

- `uv lock --check` — passed with 217 packages resolved.
- Core package and optional detection-stack imports — passed on Python 3.12; the detection import also passed under Python 3.10.
- `pre-commit run --all-files` — all applicable hooks passed, including Ruff, formatting, codespell, and mypy.
- Targeted tests were run after each functional/docs backport and after dependency groups; all passed.
- Targeted Optuna 5 tuning tests — passed.
- Non-integration suite — **1218 passed, 5 skipped, 14 deselected**.
- Integration suite — **14 passed, 2 skipped, 1221 deselected**.
- MkDocs build under Python 3.10 — passed; canonical URL and generated poster references were inspected.
- Poster verification — 17 generated poster assets and 20 rendered `poster=` references; representative pages were inspected through local `mkdocs serve`.
- Isolated Python 3.10 source/wheel build — passed.
- `twine check --strict` — passed for both the wheel and source distribution.
- `git diff --check`, `git diff --exit-code`, and final worktree check — clean.

### GitHub Actions

At head `d930dcd`, all 40 check runs completed: **38 succeeded** and **2 release-only jobs were expectedly skipped** (`Upload GitHub Release Artifacts` and `Publish Release Package`). The successful checks include the Linux/macOS/Windows Python 3.10–3.13 matrix, integration validation, mask-extra smoke test, docs build, package build, pre-release packaging, and security checks.

## Known non-blocking warnings

No warning category caused a test, docs, import, or packaging failure, and the same recurring categories were observed before the final release cut:

| Warning                                                                                     | Why it is non-blocking here                                                                                                                                   |
| ------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| NumPy divide-by-zero, overflow, or invalid-matmul warnings in Kalman/integration edge cases | Exercised by numerical edge-case/regression inputs; assertions and all integration baselines pass.                                                            |
| `invalid value encountered in sqrt` in converter tests                                      | The tests deliberately pass a negative scale to verify invalid-input behavior.                                                                                |
| McByte timestamp/mask-pipeline warning                                                      | Intentional product warning explaining that masks advance per call while timestamp-based Kalman state advances by elapsed time.                               |
| `apply_cmc_batch` `FutureWarning`                                                           | Existing deprecation notice for removal in 3.0, not a 2.6.1 failure.                                                                                          |
| Material for MkDocs 2.0 notice                                                              | Upstream informational warning; the pinned MkDocs build succeeds.                                                                                             |
| Existing `cbiou.md#buffer-ordering` anchor notice                                           | Pre-existing documentation link notice, observed before these final changes; docs build succeeds.                                                             |
| `python-doctr==1.0.1` has no `torch` extra                                                  | Upstream dependency-metadata warning during resolution; the lock, optional imports, and full suites pass.                                                     |
| Duplicate PyAV/OpenCV `AVFFrameReceiver`/`AVFAudioReceiver` classes on macOS                | Platform-specific binary-bundling warning from the optional detection stack; imports and tests complete without a crash, and hosted cross-platform CI passes. |
| Local `/versions.json` 404 while serving docs                                               | Expected outside a Mike deployment; the deployed version index is not created by plain local `mkdocs serve`.                                                  |

## Release-preparation status

- The PR remains a draft and has not created a tag or production release.
- Local artifacts correctly still use version `2.6.0` because release preparation has not run.
- After scope approval, run `/release prepare 2.6.1`, review and apply the generated changelog/version/lock changes, and rerun the final verification gates before marking the PR ready.

### CODE REVIEW

**Comment by tstanczyk95** [comment]
@Borda I updated the PR and its description with the complete 2.6.1 scope, including adapted backports, exclusions, and validation results. The branch is frozen through #599 and all local and GitHub checks are green. 

I can run `/release prepare 2.6.1` after approval.

**Review by Borda — APPROVED** [review]
LGTM, let's add the created release notes as comments when you have it.
Alos when you paste the commit SHA, do not wrap it with ` ` so GH automaticly link it via href to the particular commit

**Comment by tstanczyk95** [comment]
# Trackers 2.6.1 — release notes

## Summary

Trackers 2.6.1 fixes tracker output and mask lifecycle edge cases, restores fractional coordinates in integer-input box conversion, and reads numbered image frames in temporal order. It also prevents malformed boxes from receiving a spurious GIoU bonus and repairs URLs generated by versioned documentation deployment. The patch updates the optional detection dependency floor and refreshes the lockfile and release tooling through #599; users of OC-SORT output, integer box conversion, or the detection extra should review the migration guidance below.

## Highlights

- **OC-SORT returns every supplied detection:** a detection below `high_conf_det_threshold` now appears in `update()` output with `tracker_id=-1` instead of disappearing. The row remains in the returned detections, though output order can differ from input order ([#566](https://github.com/roboflow/trackers/pull/566)).
- **McByte delivers mask removals after interruptions:** removal notifications remain queued across skipped frames, duplicate timestamps, and recoverable CUDA out-of-memory failures, including tracks pruned early during timestamp updates ([#568](https://github.com/roboflow/trackers/pull/568)).
- **Input and scoring edge cases behave predictably:** image files such as `2.jpg` precede `10.jpg`, integer box decoding retains fractional coordinates, and inverted boxes cannot gain an invalid positive GIoU bonus ([#558](https://github.com/roboflow/trackers/pull/558), [#557](https://github.com/roboflow/trackers/pull/557), [#589](https://github.com/roboflow/trackers/pull/589)).

## Migration guide

- **OC-SORT output:** if downstream code assumed low-confidence detections were absent, filter detections before `update()` when `confidence` is available, or filter the returned rows by confidence. Low-confidence rows have `tracker_id=-1`; detections without confidence remain eligible for association regardless of the threshold. Do not rely on output row order matching input order.
- **Integer-input box conversion:** `xcycsr_to_xyxy` now promotes integer arrays to `float64` before arithmetic for single boxes and batches, preserving fractional coordinates and avoiding integer-product overflow. Code requiring integer boxes should round or cast the decoded array explicitly; floating-point input retains its dtype, including `float16`.
- **Detection extra:** `trackers[detection]` now requires `inference-models>=0.36.0`; update an older installed detection stack when upgrading. The final 2.6.1 lock resolves it to `0.37.0`.

No other public API removal or new deprecation is present in this release range.

## Notable changes

### Fixed

- **McByte and Kalman state:** mask-removal delivery is durable, while BoT-SORT and McByte refresh process noise for prediction and measurement noise for update instead of rebuilding the unused matrix ([#568](https://github.com/roboflow/trackers/pull/568), [#567](https://github.com/roboflow/trackers/pull/567)).
- **OC-SORT output, box conversion, image input, and GIoU:** the output and numeric corrections described above are covered by regression tests ([#566](https://github.com/roboflow/trackers/pull/566), [#557](https://github.com/roboflow/trackers/pull/557), [#558](https://github.com/roboflow/trackers/pull/558), [#589](https://github.com/roboflow/trackers/pull/589)).
- **Versioned documentation URLs:** Mike-deployed canonical, Open Graph, structured-data, and sitemap URLs target the `/latest/` alias instead of `/latest/2.6.0/` ([#582](https://github.com/roboflow/trackers/pull/582)).

### Changed

- **Optional detection dependency:** raised the `inference-models` minimum from `0.19.0` to `0.36.0` for the detection extra ([#592](https://github.com/roboflow/trackers/pull/592)).
- **Dependency lock:** refreshed the applicable 2.6 dependency graph. The final lock contains `cryptography 50.0.0`, `hydra-core 1.3.4`, `optuna 5.0.0`, `transformers 5.15.1`, and `uv 0.12.13`. These are dependency updates, with no specific security fix claimed ([#537](https://github.com/roboflow/trackers/pull/537), [#542](https://github.com/roboflow/trackers/pull/542), [#546](https://github.com/roboflow/trackers/pull/546), [#570](https://github.com/roboflow/trackers/pull/570), [#575](https://github.com/roboflow/trackers/pull/575), [#581](https://github.com/roboflow/trackers/pull/581), [#588](https://github.com/roboflow/trackers/pull/588), [#591](https://github.com/roboflow/trackers/pull/591), [#592](https://github.com/roboflow/trackers/pull/592), [#599](https://github.com/roboflow/trackers/pull/599)).
- **Release tooling:** updated GitHub Actions and pre-commit pins, moved pre-commit autoupdates to monthly, and ignored local `benchmarks/` output ([#544](https://github.com/roboflow/trackers/pull/544), [#550](https://github.com/roboflow/trackers/pull/550), [#590](https://github.com/roboflow/trackers/pull/590), [#593](https://github.com/roboflow/trackers/pull/593), 1770d65929dc6bbab9639ddc3301104405610c59, 330a9ee42a32e40a7b9b10cdf4a89eec32d5b8fe).

### Documentation

- Synchronized McByte benchmark and tracker comparison tables, tracker defaults, quickstart and evaluation guidance, dataset details, README terminology, search-result titles, and video poster frames with the 2.6 release layout ([#539](https://github.com/roboflow/trackers/pull/539), [#547](https://github.com/roboflow/trackers/pull/547), [#555](https://github.com/roboflow/trackers/pull/555), [#564](https://github.com/roboflow/trackers/pull/564), [#583](https://github.com/roboflow/trackers/pull/583), [#585](https://github.com/roboflow/trackers/pull/585), [#587](https://github.com/roboflow/trackers/pull/587)).

## Contributors

- **Jesús Royeth** (@JESUSROYETH) — fixed OC-SORT low-confidence output and McByte mask cleanup; corrected Kalman noise refresh.
- **Alexander Bodner** (@AlexBodner) — fixed box conversion, image-directory ordering, and documentation canonical URLs; improved release documentation.
- **Mahathir Mohammad Shuvo** (@MahathirMohammadShuvo) — fixed GIoU scoring for malformed boxes.
- **Jirka Borovec** (@Borda) — corrected tracker documentation and McByte summaries; updated pre-commit configuration and contributed to tracker fixes.
- **Tomasz Stanczyk** (@tstanczyk95) — corrected converter dtype handling after review and adapted release-branch documentation, dependency lock updates, and pre-commit changes.

*Automated contributions: @claude[bot], @dependabot[bot], @pre-commit-ci[bot], Claude Fable 5, Claude Opus 5, Copilot Autofix powered by AI, Cursor, OpenAI Codex*

**Full changelog:** [2.6.0 to current release head](https://github.com/roboflow/trackers/compare/2.6.0...463877318d7b659f612ec356416f35293ebb8843). The release tag is not created yet.

---

**PR review notes:** Contributor credits above are based on local Git authors and co-authors; PR-author attribution was not independently verified and should be confirmed before publication. On the current head, the non-integration suite passed (1,271 passed, 3 skipped, 14 deselected, 1,086 warnings), and all 14 selected integration tests passed (1,274 deselected, 48 warnings). The 2.6.1 wheel and source distribution built under Python 3.10 and passed `twine check --strict`. All six GitHub Actions workflows completed successfully on current PR head 463877318d7b659f612ec356416f35293ebb8843, which includes the changelog and converter-dtype follow-ups after release preparation; the production release jobs and pre-release PyPI upload step were skipped on this PR.

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,10 +6,26 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
...
+- **Dependency lock and release tooling refreshed through #599** — the final lock resolves `cryptography` to `50.0.0`, `hydra-core` to `1.3.4`, `optuna` to `5.0.0`, `transformers` to `5.15.1`, and `uv` to `0.12.13`; the GitHub Actions, Ruff, mypy, and pre-commit pins were also updated. These are dependency/tooling updates, with no claimed security fix ([#537](https://github.com/roboflow/trackers/pull/537), [#542](https://github.com/roboflow/trackers/pull/542), [#544](https://github.com/roboflow/trackers/pull/544), [#546](https://github.com/roboflow/trackers/pull/546), [#550](https://github.com/roboflow/trackers/pull/550), [#570](https://github.com/roboflow/trackers/pull/570), [#575](https://github.com/roboflow/trackers/pull/575), [#581](https://github.com/roboflow/trackers/pull/581), [#588](https://github.com/roboflow/trackers/pull/588), [#590](https://github.com/roboflow/trackers/pull/590), [#591](https://github.com/roboflow/trackers/pull/591), [#593](https://github.com/roboflow/trackers/pull/593), [#599](https://github.com/roboflow/trackers/pull/599)).
```
Comment (Borda) [inline CHANGELOG.md]: This is not needed; it is not shipped to the user

File: src/trackers/utils/converters.py [hunk 2]
```diff
@@ -182,7 +185,8 @@ def xcycsr_to_xyxy(xcycsr: np.ndarray) -> np.ndarray:
     # Inner np.where substitutes 1.0 for zero denominators to suppress the
     # eager-evaluation divide-by-zero warning; outer np.where replaces those results with 0.0.
     h = np.where(w != 0, xcycsr[:, 2] / np.where(w != 0, w, 1.0), 0.0)
-    result = np.empty((xcycsr.shape[0], 4), dtype=xcycsr.dtype)
+    # Decoded values are floats, so promote integers rather than truncate them; float input keeps its precision.
+    result = np.empty((xcycsr.shape[0], 4), dtype=np.result_type(xcycsr.dtype, np.float32))
```
[the remark this answers is not in the document]
Comment (tstanczyk95) [reply src/trackers/utils/converters.py]: Resolved in 4638773.
